### PR TITLE
Specialize SegmentPostings::skip_next

### PR DESCRIPTION
**Pull request by lnicola**
Old:
```
test postings::tests::bench_skip_next_p01                  ... bench:     455,790 ns/iter (+/- 91,715)
test postings::tests::bench_skip_next_p1                   ... bench:     585,395 ns/iter (+/- 91,711)
test postings::tests::bench_skip_next_p10                  ... bench:     838,795 ns/iter (+/- 160,703)
test postings::tests::bench_skip_next_p90                  ... bench:     463,535 ns/iter (+/- 77,061)
```

New:
```
test postings::tests::bench_skip_next_p01                  ... bench:      98,972 ns/iter (+/- 8,994)
test postings::tests::bench_skip_next_p1                   ... bench:     402,785 ns/iter (+/- 256,938)
test postings::tests::bench_skip_next_p10                  ... bench:     816,183 ns/iter (+/- 124,335)
test postings::tests::bench_skip_next_p90                  ... bench:     444,187 ns/iter (+/- 27,480)
```

New with exponential search:
```
test postings::tests::bench_skip_next_p01                  ... bench:      98,518 ns/iter (+/- 2,588)
test postings::tests::bench_skip_next_p1                   ... bench:     349,363 ns/iter (+/- 22,189)
test postings::tests::bench_skip_next_p10                  ... bench:     694,869 ns/iter (+/- 18,959)
test postings::tests::bench_skip_next_p90                  ... bench:     475,763 ns/iter (+/- 41,806)
```

### CODE REVIEW

File: src/postings/segment_postings.rs [hunk 1]
```diff
@@ -14,6 +14,7 @@ const EMPTY_DATA: [u8; 0] = [0u8; 0];
 /// Positions on the other hand, are optionally entirely decoded upfront.
 pub struct SegmentPostings<'a> {
     len: usize,
+    block_len: usize,
```
Comment (fulmicoton) [inline src/postings/segment_postings.rs]: can't we get the block_len from the `BlockDecoder` ?

File: src/postings/segment_postings.rs [hunk 2]
```diff
@@ -98,6 +103,81 @@ impl<'a> DocSet for SegmentPostings<'a> {
...
+        loop {
+            // check if we need to go to the next block
+            if target > self.block_decoder.output(self.block_len - 1) {
+                self.cur += Wrapping(self.block_len - pos);
+                self.load_next_block();
+                pos = 0;
+
+                // there was no more data
+                if self.block_len == 0 {
+                    return SkipResult::End;
```
Comment (fulmicoton) [inline src/postings/segment_postings.rs]: I need to double check that BlockDecoder really has that kind of contracts.

Alternatively a comparison of `self.len` and `self.cur.0`

File: src/postings/segment_postings.rs [hunk 3]
```diff
@@ -98,6 +103,81 @@ impl<'a> DocSet for SegmentPostings<'a> {
...
+        // we're in the right block, do a binary search
+        let mut start = pos;
+        let mut count = self.block_len - start;
+        while count > 0 {
+            let step = count / 2;
+            let mid = start + step;
+            let doc = self.block_decoder.output(mid);
+            if doc < target {
+                start = mid + 1;
+                count -= step + 1;
```
Comment (fulmicoton) [inline src/postings/segment_postings.rs]: maybe add a `debug_assert!` with the invariant `count == self.block_len - start`

File: src/postings/segment_postings.rs [hunk 4]
```diff
@@ -98,6 +103,81 @@ impl<'a> DocSet for SegmentPostings<'a> {
...
+            } else if doc > target {
+                count = step;
+            } else {
+                self.cur += Wrapping(mid - pos);
```
Comment (fulmicoton) [inline src/postings/segment_postings.rs]: or we could add a 

           fn set_within_block(&mut self, inner_pos: usize);

which would be implemented as (Wrapping aside)

    `self.cur = self.cur & (NUM_DOCS_PER_BLOCK - 1) + inner_pos`

File: src/postings/segment_postings.rs [hunk 5]
```diff
@@ -98,6 +103,81 @@ impl<'a> DocSet for SegmentPostings<'a> {
...
+                break;
+            }
+        }
+
+        assert!(target >= self.block_decoder.output(pos));
+        assert!(target <= self.block_decoder.output(self.block_len - 1));
+
+        // we're in the right block, do a binary search
+        let mut start = pos;
+        let mut count = self.block_len - start;
```
Comment (fulmicoton) [inline src/postings/segment_postings.rs]: should we call `count` `interval` ?

File: src/postings/segment_postings.rs [hunk 6]
```diff
@@ -98,6 +103,81 @@ impl<'a> DocSet for SegmentPostings<'a> {
...
+                if self.advance() {
+                    return SkipResult::OverStep;
+                }
+                return SkipResult::End;
+            }
+        }
+
+        // `doc` must be >= `target`
+        let doc = self.block_decoder.output(start);
+        self.cur += Wrapping(start - pos);
```
Comment (fulmicoton) [inline src/postings/segment_postings.rs]: same comment as above,

`self.set_within_block(start)`

**Comment by fulmicoton** [comment]
LGTM.

I really appreciated the super meaningful comments.

You posted a change on the segment postings performance. I didn't expect anything to change there. This is weird. Did moving `block_len` in the block postings give an accidental win on perf ?
Note that it is not exactly where it should be yet. See #89 for more info.

Could you had a benchmark for the usage of skip. You can reuse what is done in segment postings, and skip different length. It would be awesome for instance, to know, how many elements one needs to skip in average to have the current implementation of `skipping` beat the previous one.

**Comment by lnicola** [comment]
I ended up removing the early exit since it made things slower overall.

I would have liked to move the search code to a different function, but it was slower, regardless of me trying to use `get_unchecked` and/or `#[inline]`. The `std` binary search is also much slower.

I implemented the exponential search and it seems faster, but we might want to revisit this once `SegmentPostings` becomes rewindable, as the timings are much too noisy for me.

**Comment by fulmicoton** [comment]
@lnicola good job, is there other stuff you want to test or are we good here?

**Comment by fulmicoton** [comment]
on my laptop.

**After**
```
test postings::tests::bench_skip_next_p01                  ... bench:     105,033 ns/iter (+/- 18,449)
test postings::tests::bench_skip_next_p1                   ... bench:     410,501 ns/iter (+/- 188,668)
test postings::tests::bench_skip_next_p10                  ... bench:     930,328 ns/iter (+/- 176,705)
test postings::tests::bench_skip_next_p90                  ... bench:     579,244 ns/iter (+/- 79,471)
```

**Before**
```
test postings::tests::bench_skip_next_p01                  ... bench:     652,467 ns/iter (+/- 210,771)
test postings::tests::bench_skip_next_p1                   ... bench:     773,190 ns/iter (+/- 182,173)
test postings::tests::bench_skip_next_p10                  ... bench:   1,018,934 ns/iter (+/- 183,300)
test postings::tests::bench_skip_next_p90                  ... bench:     555,098 ns/iter (+/- 159,895)
```

**Comment by lnicola** [comment]
I think we're good.

**Comment by fulmicoton** [comment]
Experimenting with replacing the complete code the exponential search + the binary search by 
on optimized branchless version
`
test postings::tests::bench_skip_next_p01                  ... bench:      68,663 ns/iter (+/- 11,089)
test postings::tests::bench_skip_next_p1                   ... bench:     285,002 ns/iter (+/- 54,108)
test postings::tests::bench_skip_next_p10                  ... bench:   1,250,623 ns/iter (+/- 139,256)
test postings::tests::bench_skip_next_p90                  ... bench:     612,974 ns/iter (+/- 86,940)`

**Pull request merged**